File: hungry_frog/java/hungry_frog/src/main/java/org/example/models/Pond.java
package org.example.models;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Random;

public class Pond {

    public HashMap<Integer, LillyPad> lillyPadMap;

    public Pond(List<String> paths) {
        this.lillyPadMap = new HashMap<>();

        for (String path : paths) {
            String[] pathArray = path.split(",");
            Integer id = Integer.parseInt(pathArray[0]);
            this.lillyPadMap.put(id, new LillyPad(id, false, new ArrayList<>()));
        }

        for (String path : paths) {
            String[] pathArray = path.split(",");
            Integer id = Integer.parseInt(pathArray[0]);
            LillyPad lillyPad = this.lillyPadMap.get(id);
            for (int i = 1; i < pathArray.length; i++) {
                Integer pathId = Integer.parseInt(pathArray[i]);
                LillyPad pathLillyPad = this.lillyPadMap.get(pathId);
                lillyPad.paths.add(pathLillyPad);
            }
        }

    }

    public LillyPad placeFood(Integer foodLocationOverride) {
        LillyPad foodLocation = this.randomLillyPad(foodLocationOverride);
        foodLocation.isThereFood = true;
        return foodLocation;
    }

    public LillyPad birth(Integer birthPlaceOverride) {
        while (true) {
            LillyPad birthPlace = this.randomLillyPad(birthPlaceOverride);
            if (!birthPlace.isThereFood) {
                return birthPlace;
            }
        }
    }

    private LillyPad randomLillyPad(Integer id) {
        if (id != null) {
            return this.lillyPadMap.get(id);
        }
        Random rand = new Random();
        int randomIndex = rand.nextInt(this.lillyPadMap.size()-1) + 1;
        return this.lillyPadMap.get(randomIndex);
    }
}

File: hungry_frog/go/starter/model/pond.go
package model

import (
	"fmt"
	"math/rand"
	"strconv"
	"strings"
)

// These are for manual testing. Set to 0 to disable
var foodLocationOverride = 0
var birthPlaceOverride = 0

type Pond struct {
	LillyPads map[int]*LillyPad
}

func NewPond(paths []string) Pond {
	pond := Pond{
		map[int]*LillyPad{},
	}

	// Create the lilly pads
	for _, path := range paths {
		s := strings.Split(path, ",")
		i, _ := strconv.Atoi(s[0])
		pond.LillyPads[i] = &LillyPad{Id: i}
	}

	// Add neighbors
	for _, path := range paths {
		s := strings.Split(path, ",")
		l, _ := strconv.Atoi(s[0])
		for i := 1; i < len(s); i++ {
			n, _ := strconv.Atoi(s[i])
			pond.LillyPads[l].Paths = append(pond.LillyPads[l].Paths, pond.LillyPads[n])
		}
	}

	// Place food
	foodLocation := pond.foodLocation()
	fmt.Printf("Food location: %d\n", foodLocation.Id)
	foodLocation.IsThereFood = true

	return pond
}

func (pond *Pond) Birth() *LillyPad {
	for {
		lillyPad := pond.birthPlace()

		if !lillyPad.IsThereFood {
			return lillyPad
		}
	}
}

func (pond *Pond) randomLillyPad() *LillyPad {
	return pond.LillyPads[rand.Intn(len(pond.LillyPads)-1)+1]
}

func (pond *Pond) foodLocation() *LillyPad {
	if foodLocationOverride > 0 {
		return pond.LillyPads[foodLocationOverride]
	} else {
		return pond.randomLillyPad()
	}
}

func (pond *Pond) birthPlace() *LillyPad {
	if birthPlaceOverride > 0 {
		return pond.LillyPads[birthPlaceOverride]
	} else {
		return pond.randomLillyPad()
	}
}
